espnet2/enh/loss/criterions/tf_domain.py

```diff
@@ -0,0 +1,192 @@
+            mask = mask.int()
+        elif mask_type == "IRM":
+            # TODO(Wangyou): need to fix this,
+            #  as noise referecens are provided separately
+            mask = abs(r) / (sum(([abs(n) for n in ref_spec])) + EPS)
+        elif mask_type == "IAM":
+            mask = abs(r) / (abs(mix_spec) + EPS)
+            mask = mask.clamp(min=0, max=1)
+        elif mask_type == "PSM" or mask_type == "NPSM":
+            phase_r = r / (abs(r) + EPS)
+            phase_mix = mix_spec / (abs(mix_spec) + EPS)
+            # cos(a - b) = cos(a)*cos(b) + sin(a)*sin(b)
+            cos_theta = phase_r.real * phase_mix.real + phase_r.imag * phase_mix.imag
+            mask = (abs(r) / (abs(mix_spec) + EPS)) * cos_theta
+            mask = (
+                mask.clamp(min=0, max=1)
+                if mask_type == "NPSM"
+                else mask.clamp(min=-1, max=1)
+            )
+        elif mask_type == "PSM^2":
+            # This is for training beamforming masks
+            phase_r = r / (abs(r) + EPS)
+            phase_mix = mix_spec / (abs(mix_spec) + EPS)
+            # cos(a - b) = cos(a)*cos(b) + sin(a)*sin(b)
+            cos_theta = phase_r.real * phase_mix.real + phase_r.imag * phase_mix.imag
+            mask = (abs(r).pow(2) / (abs(mix_spec).pow(2) + EPS)) * cos_theta
+            mask = mask.clamp(min=-1, max=1)
+        assert mask is not None, f"mask type {mask_type} not supported"
+        mask_label.append(mask)
+    return mask_label
+
+
+class FrequencyDomainLoss(AbsEnhLoss, ABC):
+
+    # The loss will be computed on mask or on spectrum
+    @property
+    @abstractmethod
+    def compute_on_mask() -> bool:
+        pass
+
+    # the mask type
+    @property
+    @abstractmethod
+    def mask_type() -> str:
+        pass
+
+    def create_mask_label(self, mix_spec, ref_spec):
+        return _create_mask_label(
+            mix_spec=mix_spec, ref_spec=ref_spec, mask_type=self.mask_type
+        )
+
+
+class FrequencyDomainMSE(FrequencyDomainLoss):
+    def __init__(self, compute_on_mask=False, mask_type="IBM"):
+        super().__init__()
+        self._compute_on_mask = compute_on_mask
+        self._mask_type = mask_type
+
+    @property
+    def compute_on_mask(self) -> bool:
+        return self._compute_on_mask
+
+    @property
+    def mask_type(self) -> str:
+        return self._mask_type
+
+    @property
+    def name(self) -> str:
+        if self.compute_on_mask:
+            return f"MSE_on_{self.mask_type}"
+        else:
+            return "MSE_on_Spec"
+
+    def forward(self, ref, inf) -> torch.Tensor:
+        """time-frequency MSE loss.
+
+        Args:
+            ref: (Batch, T, F) or (Batch, T, C, F)
+            inf: (Batch, T, F) or (Batch, T, C, F)
+        Returns:
+            loss: (Batch,)
+        """
+        assert ref.shape == inf.shape, (ref.shape, inf.shape)
+        if not is_torch_1_3_plus:
+            # in case of binary masks
+            ref = ref.type(inf.dtype)
```

Comment:
We don't need this check any more. (see https://github.com/espnet/espnet/pull/3355#discussion_r679649166)

Reply:
Done.